### PR TITLE
issue 584 email consent mandatory

### PR DESCRIPTION
**Pull request by baob**
completes #584

### CODE REVIEW

File: app/views/users/show/_confirm_incoming_swap.html.haml [hunk 1]
```diff
@@ -29,13 +29,13 @@
...
-                  name: "swap[dummy]", value: "dummy"}
+                  name: "swap[dummy]", value: "dummy" }
           %input.form-check-input{ type: "checkbox", id: "consent-check",
                                    name: "swap[consent_share_email_chosen]" }
           %label.form-check-label{ for: "consent-check" }
-            Share my email address with #{user.swapped_with.redacted_name}
+            I understand that my email address will be shared with #{user.swapped_with.redacted_name}
 
-        %p.text-center{style: "width: 100%"}
+        %p.text-center{ style: "width: 100%" }
```
[the remark this answers is not in the document]
Comment (baob) [reply app/views/users/show/_confirm_incoming_swap.html.haml:38]: Later, hound, later

File: app/views/users/show/_confirm_outgoing_swap.html.haml [hunk 2]
```diff
@@ -35,20 +34,19 @@
...
+          name: "swap[dummy]", value: "dummy" }
         %input.form-check-input{ type: "checkbox", id: "swap_consent_share_email_chooser",
-          name: "swap[consent_share_email_chooser]", value: "true"}
+          name: "swap[consent_share_email_chooser]", value: "true" }
         %label.form-check-label{ for: "swap_consent_share_email_chooser" }
-          Share my email address with #{user.swapped_with.redacted_name}
+          I understand that my email address will be shared with #{user.swapped_with.redacted_name}
 
-      %p.text-center{style: "width: 100%"}
+      %p.text-center{ style: "width: 100%" }
```
[the remark this answers is not in the document]
Comment (baob) [reply app/views/users/show/_confirm_outgoing_swap.html.haml:43]: Later, hound, later

**Comment by aspiers** [comment]
Rebased

**Comment by baob** [comment]
Rebased again, so this time we should see a preview app on heroku.

**Comment by aspiers** [comment]
Rebased to be safe.

**Review by tdg — APPROVED** [review]
This works, well done. 

In an ideal world the page would not reload unless consent was given, with a tooltip message, in the same way as we have for consent to processing personal data - is there any code we can reuse from #237? 

The danger with the current way is that it ultimately takes the user back to the start of choosing a swap partner. Whereas a tooltip would keep the user in their current flow.

That said, this appears to work in the primary signup flow and should be merged.
Screenshot 2022-06-14 at 11 45 16
Screenshot 2022-06-14 at 11 45 04
Screenshot 2022-06-14 at 11 43 14

**Comment by tdg** [comment]
Am I right in thinking that as it stands, a user for whom there are no swaps available yet, is not asked to share their email address with future prospective partners, but only when a prospective swapper selects them as a swap?

**Comment by baob** [comment]
> Am I right in thinking that as it stands, a user for whom there are no swaps available yet, is not asked to share their email address with future prospective partners, but only when a prospective swapper selects them as a swap?

That appears to be right yes. We only record consent against a swap (with a known swapper) not generically.

**Comment by tdg** [comment]
>exactly how to reproduce that it would help
Not sure, I was in the process of creating a second user with the email address tom+swapmyvote+dev2@domain
In chrome

I'll see if I get it again now… no, I'm through, that's a relief… more soon

**Comment by aspiers** [comment]
[**@baob**](https://github.com/baob) commented on [June 14, 2022 1:25 PM](https://github.com/swapmyvote/swapmyvote/pull/596#issuecomment-1155114721 "2022-06-14T12:25:19Z - Replied by Github Reply Comments"):
> > Am I right in thinking that as it stands, a user for whom there are no swaps available yet, is not asked to share their email address with future prospective partners, but only when a prospective swapper selects them as a swap?
> 
> That appears to be right yes. We only record consent against a swap (with a known swapper) not generically. 
> 
> @aspiers any comment ☝️ ?

That's exactly what #345 is about changing.

**Comment by aspiers** [comment]
[**@baob**](https://github.com/baob) commented on [June 14, 2022 1:27 PM](https://github.com/swapmyvote/swapmyvote/pull/596#issuecomment-1155116751 "2022-06-14T12:27:12Z - Replied by Github Reply Comments"):
> > I've just tried to test the 'incoming swap' flow using a second account, but encountered the error page
> 
> Errr ... looks like a show-stopper ... if you can tell me exactly how to reproduce that it would help ... I haven't yet discovered where to read the actual error message that generated. Any ideas @aspiers ?

It's in Airbrake.  Access it through Heroku.

**Comment by tdg** [comment]
Screenshot 2022-06-14 at 15 04 20
Screenshot 2022-06-14 at 15 04 07
Screenshot 2022-06-14 at 15 04 00
It works on incoming swaps

**Pull request merged**